src/mp/snapshot.h

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2020 The reone project contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <cstdint>
+#include <set>
+#include <map>
+#include <list>
+#include <memory>
+#include <vector>
+#include <deque>
+
+#include "objects.h"
+#include "../common/streamreader.h"
+#include "../common/streamwriter.h"
+
+namespace reone {
+
+namespace net {
+
+constexpr uint32_t g_ServerMaxLen = 64;
+constexpr uint32_t g_ClientMaxLen = 3;  // may increase if too tight
+
+struct Snapshot {
+    Snapshot(uint32_t seq, uint32_t tick) : id_table{}, seq{ seq }, tick{ tick } {};
+
+    void reset(uint32_t newSeq, uint32_t newTick);
+
+    Snapshot &operator=(const Snapshot &) = delete;
+    //Snapshot &operator=(Snapshot&&) = delete;
+
+
+    std::map<uint32_t, std::unique_ptr<BaseStatus>> id_table;
+    uint32_t seq;
+    uint32_t tick;
+};
+
+/* 
+* memory-efficient (cough cough) snapshot system
+* sort of like a skip list */
+class History {
+public:
+    History(uint32_t maxLen) : _snapshots{}, _top{}, _seq{ 0 }, _maxLen(maxLen), _activeObjects{} {
```

Comment:
No need to explicitly initialize STL variables, i.e. `_snapshots` and `_activeObjects`. Primitive variables I would initialize in-place, e.g. `uint32_t _seq { 0 }`.